### ISSUE TITLE
Make customized runtime to test specific behavior of Polkadot Host implementations

### ISSUE THREAD

**Issue by drskalman** [w3f/polkadot-spec#145]
Similar to ethereum cross client tests each node is able to ingest and start a test with specific behavoir we should be able to make customized runtime which tells the client to behave in a specific way that we want and communicate with other clients. This also related to #140 as customized runtime allows us to make customized genesis state. As the first step we are going to run Kagome with a custome runtime and genesis and document the step. The we would like to simulate the tests which is done by ethereum cross-client tests.

**Comment by drskalman** [comment]
- @FlorianFranzen starts with the simplest runtime. Using perhaps the system module, make a genesis state of 1 or few keys, read
the genesis states and write the results in the state and print out the state tree hash.

**Comment by FlorianFranzen** [comment]
Latest implementation: https://gitlab.w3f.tech/florian/wasm-tester/

**Comment by drskalman** [comment]
both initiatial_block will print the root. The first version to be included in the new CI.

**Comment by drskalman** [comment]
@FlorianFranzen will work on this (using a simple template runtime) after fixing the Julia scripts.

**Comment by drskalman** [comment]
@FlorianFranzen will work on this (using a simple template runtime) after fixing the Julia scripts.

**Comment by drskalman** [comment]
#179 is off @FlorianFranzen shoulder and he'll work on this one next.

**Comment by drskalman** [comment]
These runtimes are being called "Host Testers". We start with working on legacy runtime API first because our goal is to test kagome now.

**Comment by drskalman** [comment]
Review by @drskalman and then we deciede about the next Runtime to implement.

**Comment by drskalman** [comment]
It is on host-tester branch.

**Comment by drskalman** [comment]
Currently it print out every function that is called. Printing out the root. Waiting for having a julia test.

**Comment by FlorianFranzen** [comment]
I finished the refactoring and extending of the existing Julia code in 4298cec (also fixes #192), which I am now using to write the missing host-test fixture. Currently mostly misses full CI support for genesis conversion (json and to yaml) and some smaller fixes.

**Comment by drskalman** [comment]
- @florianfranzen the pull request fixed. It only check that a runtime function called at all. One can put limitation 
     on the order. 
   - @gossammar isn't working partly because they are not compatible with kusama runtime (in progress). we remove it for now.
   - @lamafab and @drskalman will review the pull request.
   - we'll merge the pull in current state and then fix the rest of points.
   - printing the state root stopped working but will fix by next week.

**Comment by drskalman** [comment]
@hndnklnc has submitted a review and Florian will update it.

**Comment by drskalman** [comment]
@drskalman managed to build and run the spec test on arch and debian. 
@FlorianFranzen is adding the implementations dockers and submodules, starting with the submodule one and the back to @drskalman

**Comment by drskalman** [comment]
Commit new Makefile to host tester branch and ping @drskalman.

**Comment by drskalman** [comment]
- [x] Shouldn't Debian package list contains jq and yq? perhapse 
      apt-get install python-pip jq
      pip install yq
 - [x] Add the bit about initing with submodule and updating them will get the implementation necessary for host-tester to the README.md

**Comment by drskalman** [comment]
second run of `make test` fails if you don't manually delete '../lib/libwasmer.so'
```
make[1]: Leaving directory '/home/user1/polkadot-spec/test/hosts'
make -C hosts gossamer
make[1]: Entering directory '/home/user1/polkadot-spec/test/hosts'
cd gossamer && go build -v ./cmd/gossamer
cp gossamer/gossamer ../bin/
cp ~/go/pkg/mod/github.com/wasmerio/go-ext-wasm@*/wasmer/libwasmer.so ../lib/
cp: cannot create regular file '../lib/libwasmer.so': Permission denied
make[1]: *** [Makefile:21: gossamer] Error 1
make[1]: Leaving directory '/home/user1/polkadot-spec/test/hosts'
```

because of this:

```
user1@willie:~/polkadot-spec/test$ ls -l lib/libwasmer.so 
-r--r--r-- 1 user1 user1 6507368 Jun 29 16:15 lib/libwasmer.so
```

**Comment by FlorianFranzen** [comment]
Implemented in #205, which was just merged into master. :tada:
